src/apis/orderbookapi.py

```diff
@@ -0,0 +1,123 @@
+"""
+OrderbookAPI for fetching relevant data using the CoW Swap Orderbook API.
+"""
+# pylint: disable=logging-fstring-interpolation
+
+from typing import Any, Optional
+import json
+import requests
+from src.helper_functions import get_logger
+from src.models import Trade, OrderExecution
+from src.constants import (
+    header,
+    REQUEST_TIMEOUT,
+    SUCCESS_CODE,
+    FAIL_CODE,
+)
+
+PROD_BASE_URL = "https://api.cow.fi/mainnet/api/v1/"
+BARN_BASE_URL = "https://barn.api.cow.fi/mainnet/api/v1/"
+
+
+class OrderbookAPI:
+    """
+    Class for fetching data from a Web3 API.
+    """
+
+    def __init__(self):
+        self.logger = get_logger()
+
+    def get_solver_competition_data(self, tx_hash: str) -> Optional[dict[str, Any]]:
+        """
+        Get solver competition data from a transaction hash.
+        The returned dict follows the schema outlined here:
+        https://api.cow.fi/docs/#/default/get_api_v1_solver_competition_by_tx_hash__tx_hash_
+        """
+        prod_endpoint_url = f"{PROD_BASE_URL}solver_competition/by_tx_hash/{tx_hash}"
+        barn_endpoint_url = f"{BARN_BASE_URL}solver_competition/by_tx_hash/{tx_hash}"
+        try:
+            json_competition_data = requests.get(
+                prod_endpoint_url,
+                headers=header,
+                timeout=REQUEST_TIMEOUT,
+            )
+            if json_competition_data.status_code == SUCCESS_CODE:
+                solver_competition_data = json.loads(json_competition_data.text)
+            elif json_competition_data.status_code == FAIL_CODE:
+                barn_competition_data = requests.get(
+                    barn_endpoint_url, headers=header, timeout=REQUEST_TIMEOUT
+                )
+                if barn_competition_data.status_code == SUCCESS_CODE:
+                    solver_competition_data = json.loads(barn_competition_data.text)
+                else:
+                    return None
+        except requests.exceptions.ConnectionError as err:
+            self.logger.warning(
+                f"Connection error while fetching competition data: {err}"
```

Comment:
Maybe it would be good to include the request that caused the exception? I am thinking mostly to include the tx hash in the error msg, in case we might need it when debugging.

Reply:
Good idea.

Reply:
Fixed.